packages/jest-types/src/Circus.ts

```diff
@@ -223,6 +229,7 @@ export type State = {
   unhandledErrors: Array<Exception>;
   includeTestLocationInResult: boolean;
   maxConcurrency: number;
+  unhandledRejectionErrorByPromise: Map<Promise<unknown>, Exception>;
```

Comment:
should this be a `WeakMap`?

Reply:
It should not to be able to get values out of it at the end of a test. If the promise gets resolved, it will be removed from the map by `rejectionHandled` handler. Otherwise, we want to be able to report the error. It could be evicted from the map using `WeakMap`.

Reply:
and a weak reference is not enough for us to achieve that?

We've hadd issues with memory leaks and this state before, which is why I'm a bit hesitant 😅 Happy to land if this is good, tho

Reply:
I totally understand. Let me elaborate a bit more about the reasoning not to use `WeakMap`:

1. I cannot iterate over values inside a weak map as its API does not support listing items, it supports direct access by keys only. Therefore, I would not be able to report all unhandled rejections at the end of a test.

2. When a promise is considered unhandled rejection, it should still live inside memory somehow so there is a chance to get `rejectionHandled` where the promise comes as parameter to the event handler.

3. If the previous assumption about keeping the promise around is wrong, I still cannot use `WeakMap` as I would lose information about unhandled rejections because `WeakMap` would release memory for key that does not live in memory anymore. Therefore, the feature would be nonsense.

However, I small improvement for that could be I remove the map completely (or replace by an empty instance) once it is processed at the end of a test. I do not think it should matter because each map and promise have pretty small memory overhead AFAIK but I can make this micro-optimisation if needed.

Reply:
if we don't manually clear e.g. `unhandledErrors` I don't think we need to clear this either

Reply:
I did not find any clearing of `unhandledErrors`.